src/memory-usage.ts

```diff
@@ -0,0 +1,109 @@
+  children: InternalTree[];
+}
+function convert(m: WireTree, title: string): InternalTree {
+  const slash = Math.max(title.lastIndexOf('/'), title.lastIndexOf('\\'));
+  return {
+    title: title.substr(slash + 1), // display basename only
+    isFile: slash >= 0,
+    total: m._total,
+    self: m._self,
+    children: Object.keys(m)
+                  .filter(x => !x.startsWith('_'))
+                  .map(e => convert(m[e] as WireTree, e))
+                  .sort((x, y) => x.total - y.total)
+                  .reverse(),
+  };
+}
+
+class ASTFeature implements vscodelc.StaticFeature {
+  constructor(private context: ClangdContext) {
+    const adapter = new TreeAdapter();
+    adapter.onDidChangeTreeData(
+        (e) => vscode.commands.executeCommand(
+            'setContext', 'clangd.memoryUsage.hasData', adapter.root != null));
+    this.context.subscriptions.push(
+        vscode.window.registerTreeDataProvider('clangd.memoryUsage', adapter));
+    vscode.commands.registerCommand('clangd.memoryUsage', async () => {
+      const usage = await this.context.client.sendRequest(RequestType, {});
+      adapter.root = convert(usage, '<root>');
+    });
+    vscode.commands.registerCommand('clangd.memoryUsage.close',
+                                    () => adapter.root = null);
+  }
+
+  fillClientCapabilities(capabilities: vscodelc.ClientCapabilities) {}
+
+  initialize(capabilities: vscodelc.ServerCapabilities,
+             _documentSelector: vscodelc.DocumentSelector|undefined) {
+    vscode.commands.executeCommand('setContext', 'clangd.memoryUsage.supported',
+                                   'memoryUsageProvider' in capabilities);
+  }
+}
+
+class TreeAdapter implements vscode.TreeDataProvider<InternalTree> {
+  private root_?: InternalTree;
+
+  get root(): InternalTree|null { return this.root_; }
+  set root(n: InternalTree|null) {
+    this.root_ = n;
+    this._onDidChangeTreeData.fire(/*root changed*/ null);
+  }
+
+  private _onDidChangeTreeData = new vscode.EventEmitter<InternalTree|null>();
+  readonly onDidChangeTreeData = this._onDidChangeTreeData.event;
+
+  public getTreeItem(node: InternalTree): vscode.TreeItem {
+    const item = new vscode.TreeItem(node.title);
+    item.description = (node.total / 1024 / 1024).toFixed(2) + ' MB';
+    item.tooltip = `self=${node.self} total=${node.total}`;
+    if (node.isFile)
+      item.iconPath = new vscode.ThemeIcon('symbol-file');
+    else if (!node.children.length)
+      item.iconPath = new vscode.ThemeIcon('circle-filled');
+    if (node.children.length) {
+      if (node.children.length >= 6 || node.isFile)
+        item.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
+      else
+        item.collapsibleState = vscode.TreeItemCollapsibleState.Expanded;
+    }
+    return item;
+  }
+
+  public async getChildren(t?: InternalTree): Promise<InternalTree[]> {
+    return t ? t.children : this.root ? [this.root] : [];
```

Comment:
i bet this reads correctly, but might be nice to have some parentheses to make it explicit :D